#**MACHINE LEARNING PROJECT ON RED WINE DATASET**

# Classification

**Data preprocessing**

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [2]:
dataset = pd.read_csv('RED WINE.csv')

In [3]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1599 entries, 0 to 1598
Data columns (total 12 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   fixed acidity         1599 non-null   float64
 1   volatile acidity      1599 non-null   float64
 2   citric acid           1599 non-null   float64
 3   residual sugar        1599 non-null   float64
 4   chlorides             1599 non-null   float64
 5   free sulfur dioxide   1599 non-null   float64
 6   total sulfur dioxide  1599 non-null   float64
 7   density               1599 non-null   float64
 8   pH                    1599 non-null   float64
 9   sulphates             1599 non-null   float64
 10  alcohol               1599 non-null   float64
 11  quality               1599 non-null   int64  
dtypes: float64(11), int64(1)
memory usage: 150.0 KB


In [4]:
dataset['quality'].unique()

array([5, 6, 7, 4, 8, 3], dtype=int64)

**Separate features and target**  


In [5]:
X = dataset.iloc[: , 1:-1].values
y = dataset.iloc[:, -1].values

**Splitting data into train and test sets**

In [6]:
from sklearn.model_selection import train_test_split

In [7]:
X_train, X_test, y_train, y_test = train_test_split(X ,y, test_size=0.20)

**Standardization**

In [8]:
from sklearn.preprocessing import StandardScaler

In [9]:
sc = StandardScaler()

In [10]:
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)

**Train on different algorithms**

In [11]:
from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC

In [12]:
l_cla = LogisticRegression()
k_cla = KNeighborsClassifier()
d_cla = DecisionTreeClassifier()
r_cla = RandomForestClassifier()
s_cla = SVC(kernel='linear')
ks_cla = SVC(kernel= 'rbf')

In [13]:
l_cla.fit(X_train, y_train)
k_cla.fit(X_train, y_train)
d_cla.fit(X_train, y_train)
r_cla.fit(X_train, y_train)
s_cla.fit(X_train, y_train)
ks_cla.fit(X_train, y_train)

SVC(C=1.0, break_ties=False, cache_size=200, class_weight=None, coef0=0.0,
    decision_function_shape='ovr', degree=3, gamma='scale', kernel='rbf',
    max_iter=-1, probability=False, random_state=None, shrinking=True,
    tol=0.001, verbose=False)

In [14]:
l_pred = l_cla.predict(X_test)
k_pred = k_cla.predict(X_test)
d_pred = d_cla.predict(X_test)
r_pred = r_cla.predict(X_test)
s_pred = s_cla.predict(X_test)
ks_pred = ks_cla.predict(X_test)

In [15]:
from sklearn.metrics import confusion_matrix

In [16]:
l_c = confusion_matrix(y_test, l_pred)
k_c = confusion_matrix(y_test, k_pred)
d_c = confusion_matrix(y_test, d_pred)
r_c = confusion_matrix(y_test, r_pred)
s_c = confusion_matrix(y_test, s_pred)
ks_c = confusion_matrix(y_test, ks_pred)

In [17]:
l_c

array([[  0,   0,   1,   0,   0,   0],
       [  0,   0,   5,   5,   1,   0],
       [  1,   0, 112,  32,   1,   0],
       [  0,   1,  34,  72,  12,   0],
       [  0,   0,   1,  26,  13,   0],
       [  0,   0,   0,   1,   2,   0]], dtype=int64)

In [18]:
k_c

array([[  0,   0,   0,   1,   0,   0],
       [  0,   0,   4,   6,   1,   0],
       [  1,   3, 100,  40,   2,   0],
       [  0,   0,  31,  70,  18,   0],
       [  0,   0,   7,  17,  16,   0],
       [  0,   0,   0,   1,   2,   0]], dtype=int64)

In [19]:
d_c

array([[  0,   0,   0,   1,   0,   0],
       [  0,   1,   3,   5,   2,   0],
       [  0,   6, 103,  35,   2,   0],
       [  0,   3,  26,  72,  15,   3],
       [  0,   1,   4,   8,  27,   0],
       [  0,   0,   0,   0,   3,   0]], dtype=int64)

In [20]:
r_c

array([[  0,   0,   0,   1,   0,   0],
       [  0,   0,   5,   6,   0,   0],
       [  0,   0, 116,  29,   1,   0],
       [  0,   0,  25,  87,   7,   0],
       [  0,   0,   3,  17,  20,   0],
       [  0,   0,   0,   2,   1,   0]], dtype=int64)

In [21]:
s_c

array([[  0,   0,   1,   0,   0,   0],
       [  0,   0,   6,   5,   0,   0],
       [  0,   0, 115,  31,   0,   0],
       [  0,   0,  38,  81,   0,   0],
       [  0,   0,   1,  39,   0,   0],
       [  0,   0,   0,   3,   0,   0]], dtype=int64)

In [22]:
ks_c

array([[  0,   0,   0,   1,   0,   0],
       [  0,   0,   6,   4,   1,   0],
       [  0,   0, 115,  31,   0,   0],
       [  0,   0,  33,  77,   9,   0],
       [  0,   0,   3,  22,  15,   0],
       [  0,   0,   0,   2,   1,   0]], dtype=int64)

In [23]:
from sklearn.metrics import accuracy_score

In [24]:
l_a = accuracy_score(y_test, l_pred)
k_a = accuracy_score(y_test, k_pred)
d_a = accuracy_score(y_test, d_pred)
r_a = accuracy_score(y_test, r_pred)
s_a = accuracy_score(y_test, s_pred)
ks_a = accuracy_score(y_test, ks_pred)

**Result**

In [25]:
print('Logistic Regression: ' + str(l_a) + '\nKNN: ' + str(k_a) + '\nDecision Tree: ' + str(d_a) + '\nRandom Forest: ' + str(r_a) + '\nLinear SVC: ' + str(s_a) + '\nKernel SVC: ' + str(l_a))

Logistic Regression: 0.615625
KNN: 0.58125
Decision Tree: 0.634375
Random Forest: 0.696875
Linear SVC: 0.6125
Kernel SVC: 0.615625


#Regression

**Data preprocessing**

In [26]:
import pandas as pd
import pandas as pd
import matplotlib.pyplot as plt

In [27]:
data=pd.read_csv("RED WINE.csv")

In [28]:
data

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0               7.4             0.700         0.00             1.9      0.076   
1               7.8             0.880         0.00             2.6      0.098   
2               7.8             0.760         0.04             2.3      0.092   
3              11.2             0.280         0.56             1.9      0.075   
4               7.4             0.700         0.00             1.9      0.076   
...             ...               ...          ...             ...        ...   
1594            6.2             0.600         0.08             2.0      0.090   
1595            5.9             0.550         0.10             2.2      0.062   
1596            6.3             0.510         0.13             2.3      0.076   
1597            5.9             0.645         0.12             2.0      0.075   
1598            6.0             0.310         0.47             3.6      0.067   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                    11.0                  34.0  0.99780  3.51       0.56   
1                    25.0                  67.0  0.99680  3.20       0.68   
2                    15.0                  54.0  0.99700  3.26       0.65   
3                    17.0                  60.0  0.99800  3.16       0.58   
4                    11.0                  34.0  0.99780  3.51       0.56   
...                   ...                   ...      ...   ...        ...   
1594                 32.0                  44.0  0.99490  3.45       0.58   
1595                 39.0                  51.0  0.99512  3.52       0.76   
1596                 29.0                  40.0  0.99574  3.42       0.75   
1597                 32.0                  44.0  0.99547  3.57       0.71   
1598                 18.0                  42.0  0.99549  3.39       0.66   

      alcohol  quality  
0         9.4        5  
1         9.8        5  
2         9.8        5  
3         9.8        6  
4         9.4        5  
...       ...      ...  
1594     10.5        5  
1595     11.2        6  
1596     11.0        6  
1597     10.2        5  
1598     11.0        6  

[1599 rows x 12 columns]

In [29]:
data.describe()

fixed acidity  volatile acidity  citric acid  residual sugar  \
count    1599.000000       1599.000000  1599.000000     1599.000000   
mean        8.319637          0.527821     0.270976        2.538806   
std         1.741096          0.179060     0.194801        1.409928   
min         4.600000          0.120000     0.000000        0.900000   
25%         7.100000          0.390000     0.090000        1.900000   
50%         7.900000          0.520000     0.260000        2.200000   
75%         9.200000          0.640000     0.420000        2.600000   
max        15.900000          1.580000     1.000000       15.500000   

         chlorides  free sulfur dioxide  total sulfur dioxide      density  \
count  1599.000000          1599.000000           1599.000000  1599.000000   
mean      0.087467            15.874922             46.467792     0.996747   
std       0.047065            10.460157             32.895324     0.001887   
min       0.012000             1.000000              6.000000     0.990070   
25%       0.070000             7.000000             22.000000     0.995600   
50%       0.079000            14.000000             38.000000     0.996750   
75%       0.090000            21.000000             62.000000     0.997835   
max       0.611000            72.000000            289.000000     1.003690   

                pH    sulphates      alcohol      quality  
count  1599.000000  1599.000000  1599.000000  1599.000000  
mean      3.311113     0.658149    10.422983     5.636023  
std       0.154386     0.169507     1.065668     0.807569  
min       2.740000     0.330000     8.400000     3.000000  
25%       3.210000     0.550000     9.500000     5.000000  
50%       3.310000     0.620000    10.200000     6.000000  
75%       3.400000     0.730000    11.100000     6.000000  
max       4.010000     2.000000    14.900000     8.000000

In [30]:
data.head()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.4              0.70         0.00             1.9      0.076   
1            7.8              0.88         0.00             2.6      0.098   
2            7.8              0.76         0.04             2.3      0.092   
3           11.2              0.28         0.56             1.9      0.075   
4            7.4              0.70         0.00             1.9      0.076   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 11.0                  34.0   0.9978  3.51       0.56   
1                 25.0                  67.0   0.9968  3.20       0.68   
2                 15.0                  54.0   0.9970  3.26       0.65   
3                 17.0                  60.0   0.9980  3.16       0.58   
4                 11.0                  34.0   0.9978  3.51       0.56   

   alcohol  quality  
0      9.4        5  
1      9.8        5  
2      9.8        5  
3      9.8        6  
4      9.4        5

In [31]:
data. info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1599 entries, 0 to 1598
Data columns (total 12 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   fixed acidity         1599 non-null   float64
 1   volatile acidity      1599 non-null   float64
 2   citric acid           1599 non-null   float64
 3   residual sugar        1599 non-null   float64
 4   chlorides             1599 non-null   float64
 5   free sulfur dioxide   1599 non-null   float64
 6   total sulfur dioxide  1599 non-null   float64
 7   density               1599 non-null   float64
 8   pH                    1599 non-null   float64
 9   sulphates             1599 non-null   float64
 10  alcohol               1599 non-null   float64
 11  quality               1599 non-null   int64  
dtypes: float64(11), int64(1)
memory usage: 150.0 KB


**Separate features and target**

In [32]:
X = data.iloc[: , :-1].values
Y = data.iloc[: , -1].values

In [33]:
X

array([[ 7.4  ,  0.7  ,  0.   , ...,  3.51 ,  0.56 ,  9.4  ],
       [ 7.8  ,  0.88 ,  0.   , ...,  3.2  ,  0.68 ,  9.8  ],
       [ 7.8  ,  0.76 ,  0.04 , ...,  3.26 ,  0.65 ,  9.8  ],
       ...,
       [ 6.3  ,  0.51 ,  0.13 , ...,  3.42 ,  0.75 , 11.   ],
       [ 5.9  ,  0.645,  0.12 , ...,  3.57 ,  0.71 , 10.2  ],
       [ 6.   ,  0.31 ,  0.47 , ...,  3.39 ,  0.66 , 11.   ]])

In [34]:
Y

array([5, 5, 5, ..., 6, 5, 6], dtype=int64)

**Splitting Data into Train and Test sets**

In [35]:
 from sklearn.model_selection import train_test_split

In [36]:
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size = 0.2)

In [37]:
X_train

array([[13.2  ,  0.46 ,  0.52 , ...,  3.1  ,  0.56 ,  9.   ],
       [ 7.4  ,  0.55 ,  0.22 , ...,  3.05 ,  0.63 ,  9.2  ],
       [ 8.3  ,  0.705,  0.12 , ...,  3.51 ,  0.72 , 10.   ],
       ...,
       [ 8.8  ,  0.64 ,  0.17 , ...,  3.23 ,  0.54 ,  9.6  ],
       [ 7.6  ,  0.35 ,  0.6  , ...,  3.38 ,  0.79 , 11.1  ],
       [ 7.6  ,  0.95 ,  0.03 , ...,  3.2  ,  0.56 ,  9.6  ]])

**Train on different algorithms**

In [38]:
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import PolynomialFeatures
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor

In [39]:
m_reg = LinearRegression()
p_reg = LinearRegression()
d_reg = DecisionTreeRegressor()
r_reg = RandomForestRegressor(n_estimators=500)

In [40]:
X_poly = PolynomialFeatures(degree =2)
X_poly = X_poly.fit_transform(X_train)

In [41]:
m_reg.fit(X_train, Y_train)
p_reg.fit(X_poly,Y_train)
d_reg.fit(X_train, Y_train)
r_reg.fit(X_train, Y_train)

RandomForestRegressor(bootstrap=True, ccp_alpha=0.0, criterion='mse',
                      max_depth=None, max_features='auto', max_leaf_nodes=None,
                      max_samples=None, min_impurity_decrease=0.0,
                      min_impurity_split=None, min_samples_leaf=1,
                      min_samples_split=2, min_weight_fraction_leaf=0.0,
                      n_estimators=500, n_jobs=None, oob_score=False,
                      random_state=None, verbose=0, warm_start=False)

In [42]:
temp = PolynomialFeatures(degree =2)
temp = temp.fit_transform(X_test)

In [43]:
m_pred = m_reg.predict(X_test)
p_pred = p_reg.predict(temp)
d_pred = d_reg.predict(X_test)
r_pred = r_reg.predict(X_test)

In [44]:
from sklearn.metrics import r2_score

In [45]:
m = r2_score(Y_test, m_pred)
p = r2_score(Y_test, p_pred)
d = r2_score(Y_test, d_pred)
r = r2_score(Y_test, r_pred)

**Result**

In [46]:
print('Logistic Regression: ' + str(m) + '\npolyfeat: ' + str(p) + '\nDecision Tree: ' + str(d) + '\nRandom Forest: ' + str(r))

Logistic Regression: 0.29896093870628315
polyfeat: 0.21834905700525553
Decision Tree: -0.27280545488052077
Random Forest: 0.3882938264021132


#**REPORT OF RED WINE DATASET**

##**1.Contents of data file**

Data file contains datasets of red_wine_quality. what are the elements needed to prepare the wine,relation between them.



##**2.Libraries imported**

i.Pandas  
ii.numpy  
iii.matplotlib

##**3.Model used**

i.Classification   
ii.Regression 

##**4.Algorithms**

1.Logistic regression  
2.Polynomial features 
3.Decision tree 
4.Random forest

##**5.Data preprocessing**

1.describing data based on mean,median,std_dev,percentile   
2.finding null values if present  
3.checking datatypes  
4.scaling data  
5.encoding mechanisms  
6.checking quality of wine

##**6.Results**

1.Logistic Regression: 0.29896093870628315
2.polynomial features: 0.21834905700525553
3.Decision Tree: -0.27280545488052077
4.Random Forest: 0.3882938264021132

##**7.Conclusion**


according to given dataset,wine quality will be good if it cross the scale above 75%.

after testing on them the purity value obtained are less than given reference value.hence, the given the quality of wine is not good..

Since Random forest is giving more efficiency than other algorithm we can consider random forest as the best algorithm for this dataset